abcd/backends/atoms_properties.py

```diff
@@ -0,0 +1,205 @@
+
+        if units is not None:
+            for key in units.keys():
+                if key not in self.df.columns.values:
+                    raise ValueError(
+                        f"Invalid field name: {key}. Keys in `units` must "
+                        "correspond to field names in the loaded data."
+                    )
+            self.units = units
+        elif infer_units:
+            self._separate_units()
+        else:
+            self.units = None
+
+        self.store_struct_file = store_struct_file
+        if self.store_struct_file:
+            if struct_file_template is not None:
+                self.struct_file_template = struct_file_template
+            else:
+                raise ValueError(
+                    "`struct_file_template` must be specified if "
+                    "store_struct_file is True."
+                )
+            if struct_name_label is not None:
+                self.struct_name_label = struct_name_label
+            else:
+                raise ValueError(
+                    "`struct_name_label` must be specified if store_struct_file is"
+                    " True."
+                )
+            self.set_struct_files()
+
+    def _separate_units(self):
+        """
+        Parse field names to determine units.
+        """
+        columns = []
+        self.units = {}
+        for column in list(self.df.columns.values):
+            if "," in column:
+                column_name = column.split(",")[0].strip()
+                self.units[column_name] = column.split(",")[1].strip()
+            elif "(" in column:
+                column_name = column.split("(")[0].strip()
+                self.units[column_name] = column.split("(")[1].strip()[:-1]
+            else:
+                column_name = column
+
+            columns.append(column_name)
+
+        self.df.columns = columns
+
+    def set_struct_files(self):
+        """
+        Sets a list containing a filename for each structure in the dataframe.
+        """
+        self.struct_files = []
+
+        for i in range(len(self.df)):
+            try:
+                struct_name = self.df.iloc[i][self.struct_name_label]
+            except KeyError:
+                raise ValueError(
+                    f"{self.struct_name_label} is not a valid column in "
+                    "the data loaded."
+                )
+            struct_file = self.get_struct_file(struct_name)
+            self.struct_files.append(struct_file)
+
+    def get_struct_file(self, struct_name: str) -> str:
+        """
+        Evaluate struct_file_template to determine structure filename
+        for current structure.
+
+        Parameters
+        ----------
+        struct_name: str
+            Name of current structure.
+
+        Returns
+        -------
+        Filename for the current structure.
+        """
+        if struct_name is None:
+            raise ValueError("`struct_name` must be specified")
```

Comment:
Is this likely? It violates the type annotation anyway. Maybe better to check against `not isinstance(.., str):`